File: 42-trapping-rain-water/solution.ts
type StackElement = { h: number; pos: number };

function trap(height: number[]): number {
  let answer = 0;
  const stack: StackElement[] = [];
  const len = height.length;

  for (let i = 0; i < len; i++) {
    let prev: StackElement = { h: -1, pos: -1 };
    while (true) {
      if (stack.slice(-1)[0]?.h <= height[i]) {
        prev = stack.pop() as StackElement;
      } else break;
    }

    if (prev.h !== -1) {
      for (let j = prev.pos; j < i; j++) {
        answer += prev.h - height[j];
        height[j] = prev.h;
      }
    }

    if (height[i] > 0) {
      stack.push({ h: height[i], pos: i });
    }
  }

  const stackLen = stack.length;
  if (stackLen > 0) {
    for (let i = 0; i < stackLen - 1; i++) {
      const h = stack[i + 1].h;
      const start = stack[i].pos;
      const end = stack[i + 1].pos;
      for (let j = start; j < end; j++) {
        answer += Math.max(h - height[j], 0);
      }
    }
  }

  return answer;
}

File: 42-trapping-rain-water/solution.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        stack<pair<int, int>> sk; // [height, idx]
        int ans = 0, n = height.size();
        for(int i = 0; i < n; i++) {
            int prev = 0;
            while(!sk.empty() && sk.top().first <= height[i]) {
                auto [h, idx] = sk.top();
                sk.pop();
                ans += (h - prev) * (i - idx - 1);
                prev = h;
            }
            if(!sk.empty()) ans += (height[i] - prev) * (i - sk.top().second - 1);
            sk.push({height[i], i});
        }
        return ans;
    }
};
